### PR TITLE
Added initial YAML ontology

### PR DESCRIPTION
**Pull request by Waguramu**
Initial YAML ontology stub to work on ヽ(o＾▽＾o)ノ

(´• ω •`)ﾉ˚ ^(●ᴥ●)^

### CODE REVIEW

**Comment by Waguramu** [comment]
Is this all?

**Comment by josephbirkner** [comment]
1. Further streamlining is possible with yaml representations (See https://pyyaml.org/wiki/PyYAMLDocumentation#constructors-representers-resolvers). Consider the the following change:
  Instead of
```yaml
Person:
   main_label: Person
   allowed_labels: [Person]
   allowed_properties: [name, sex, full_name, birthdate, timestamp]
   allowed_relationships: [EQUALS, FROM, HAS_HOBBY, LIVE_IN, STUDY_AT, OCCUPIED_AS, WORK_FOR, FRIEND_OF, MEMBER_OF, CHILD_OF, SIBLING_OF]
```
  write
```yaml
!Type
name: Person
allowed_properties: [name, sex, full_name, birthdate, timestamp]
allowed_relationships: [EQUALS, FROM, HAS_HOBBY, LIVE_IN, STUDY_AT, OCCUPIED_AS, WORK_FOR, FRIEND_OF, MEMBER_OF, CHILD_OF, SIBLING_OF]
```

2. From the current ontology file example, it looks like `allowed_labels` do not define an equivalence relationship (City|Location, Country|Location, but not Country|City). Indeed from the example, it seems like `allowed_labels` isn't just intransitive, it's not even commutative (City|Location but not Location|City). This introduces implementation uncertainties for retrieval. I would suggest removing `allowed_labels`, and introducing a `Combinable` object which represents a certain equivalence set:
```
!Combinable
[City, Location, Country]
```

**Comment by josephbirkner** [comment]
> > Country|City
> 
> You realise a country cannot be a city?

Point taken (_But Vatican City!_ 😋 ).

> City|Location is the same as Location|City because you do union of sets of their allowed properties and relations

But then, implicitly: `Location.allowed_labels={City, Country}` and the transitive property is established.

What I mean is: You could do `Node(City).add_label(Location)`, but you can not do `Node(Location).add_label(City)`.

I'm ok with that (essentially an IS-A feature?), but then ...

* how do we determine the main label upon retrieval? Should we store it in a property? It may also be possible to derive the main label procedurally as the leaf node of the IS-A graph, but that may end up with cyclic, many, or no results.
* If it essentially indicates an *IS-A*, then why not automate the process? Upon `Node(City)`, add the `Location` label automatically?

**Comment by Waguramu** [comment]
I think we have to remove add_label() method. Make them fixed and initialised upon calling Node’s constructor. We may implement type casting, so if you want Location -> City, or City -> Location, just cast it, so you can get both types and compare possibility of such things

**Comment by Waguramu** [comment]
The set of labels has to be complete, indeed - this cannot be discussed

**Comment by josephbirkner** [comment]
> The set of labels has to be complete, indeed - this cannot be discussed

I'm not following, which set are you referring to?

>  We may implement type casting, so if you want Location -> City, or City -> Location, just cast it, so you can get both types and compare possibility of such things

You mean, if you want to convert something that was previously just a Location to a City?

Please reconsider the transitive&commutative Combinable feature again wrt/ simplicity. Is it that bad if we have reptile cats? 👽

**Comment by Waguramu** [comment]
Which set?
Inside of the Node class you have labels field - it should contain all allowed_labels for the Type

**Comment by Waguramu** [comment]
No - reptile cats
Yes - type casting

**Comment by josephbirkner** [comment]
Requests:
- Drop _label_ as terminology in discussions and code, use _type_
- Rename `allowed_labels` to `is_a` or `isa`
- Rename `allowed_properties` to `properties` or `props`
- Rename `allowed_relationships` to `relationships` or `rels`

**Comment by Waguramu** [comment]
400 Bad request

**Comment by Waguramu** [comment]
See
https://neo4j.com/docs/getting-started/current/graphdb-concepts/

**Comment by josephbirkner** [comment]
Compare:
https://dev.grakn.ai/academy/schema-elements.html

**Comment by josephbirkner** [comment]
But what is the legitimazation for the unnecessarily verbose naming? If there are `allowed_things`, there must be `disallowed_things`?

**Comment by Waguramu** [comment]
But you said we don’t need disallowed things because allowed things are enough. And now you want to take this away too

**Comment by josephbirkner** [comment]
Do you like adding unnecessary prefixes? Why not call it `untyped_const_compatible_allowed_properties`? /s

File: ontology/example_ontology.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
+-------
+!Person
+type: Person
+meta: [Person]
+properties: [name, sex, full_name, birthdate, timestamp]
+relationships: [EQUALS, FROM, HAS_HOBBY, LIVE_IN, STUDY_AT, OCCUPIED_AS, WORK_FOR, FRIEND_OF, MEMBER_OF, CHILD_OF, SIBLING_OF]
+-------
+!TelegramPerson:
```
Comment (josephbirkner) [inline ontology/example_ontology.yaml]:
```suggestion
!TelegramPerson
```

File: ontology/example_ontology.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+type: TelegramPerson
+meta: [TelegramPerson]
+properties: [name, sex, full_name, birthdate, telegram_id, timestamp]
+relationships: [EQUALS, FROM, HAS_HOBBY, LIVE_IN, STUDY_AT, OCCUPIED_AS, WORK_FOR, FRIEND_OF, MEMBER_OF, CHILD_OF, SIBLING_OF]
+-------
+!Robot:
```
Comment (josephbirkner) [inline ontology/example_ontology.yaml]: Is the goal to write classes for every type? I see how `!Person` or `!Roboy` might benefit from having their own class, but most ontology types should be some `!GenericType` or so, right?

**Comment by Waguramu** [comment]
@josephbirkner Since we are changing **label -> type** due to your wish rather than necessity, it is up to you to ensure consistency everywhere: in code, in comments, in documentation, in repos, in confluence, in jira (legacy too if needed)

**Comment by josephbirkner** [comment]
> If you remove meta from there, doesn't it mean that the classes won't have this field? How do we nicely iterate over the ontology then?

Every ontology class is at least a subclass of `scientio.Type`. The yaml fields are constructor parameters, which may have default values.

**Comment by Waguramu** [comment]
Actually, nevermind

**Pull request merged**